### PR TITLE
Adjust StorageClasses volumeBindingMode

### PR DESCRIPTION
**Pull request by tuommaki**
Set `volumeBindingMode` of our `StorageClasses` to
`WaitForFirstConsumer`. This makes scheduling of dynamic volumes
topology aware as it respects Pod's scheduling constraints.

Towards https://github.com/giantswarm/giantswarm/issues/14039

### CODE REVIEW

**Review by fiunchinho — APPROVED** [review]
LGTMIWFY

**Comment by whites11** [comment]
The PR seems OK but we need to ensure the change of the field works during an upgrade.
See https://gigantic.slack.com/archives/CEYM0KY5D/p1611224419000800

File: service/controller/azureconfig/handler/volumebindingmigration/resource.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	// Iterate over StorageClasses and remove items that already have expected
+	// volume binding mode. Rest will be updated in one call.
+	for i := 0; i < len(storageClassList.Items); i++ {
+		sc := storageClassList.Items[i]
+
+		if sc.VolumeBindingMode != nil && *sc.VolumeBindingMode == storagev1.VolumeBindingWaitForFirstConsumer {
```
Comment (whites11) [inline service/controller/azureconfig/handler/volumebindingmigration/resource.go:82]: Shouldn't you use the `desiredVolumeBindMode` defined above?

File: service/controller/azureconfig/handler/volumebindingmigration/resource.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+			return microerror.Mask(err)
+		}
+	}
+
+	desiredVolumeBindMode := storagev1.VolumeBindingWaitForFirstConsumer
+
+	r.logger.Debugf(ctx, "ensuring storageclasses use desired volumeBindMode %q", desiredVolumeBindMode)
+
+	storageClassList := &storagev1.StorageClassList{}
+	err = tenantClusterK8sClient.List(ctx, storageClassList)
```
Comment (whites11) [inline service/controller/azureconfig/handler/volumebindingmigration/resource.go:72]: Maybe we should update only the storage classes we created and not all of them. There might be others created by the customer.

File: service/controller/azureconfig/handler/volumebindingmigration/resource.go (hunk 1 above)
Comment (tuommaki) [reply service/controller/azureconfig/handler/volumebindingmigration/resource.go:82]: Kind of the same thing, but yes, that would make it more readable. I'll fix :+1:

File: service/controller/azureconfig/handler/volumebindingmigration/resource.go (hunk 2 above)
Comment (tuommaki) [reply service/controller/azureconfig/handler/volumebindingmigration/resource.go:72]: Ack. I'll add a filter for those.

File: service/controller/azureconfig/handler/volumebindingmigration/resource.go [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+		}
+	}
+
+	return k8sClient.CtrlClient(), nil
+}
+
+func defaultStorageClasses() []storagev1.StorageClass {
+	var storageClasses []storagev1.StorageClass
+
+	objs := bytes.Split([]byte(ignition.DefaultStorageClass), []byte("---"))
```
Comment (tuommaki) [inline service/controller/azureconfig/handler/volumebindingmigration/resource.go:162]: I feel like this is super stupid way to do this, but there's no streaming parsing implemented in k8s YAML parser and none others works correctly with the object meta annotations :man_facepalming:

Comment (whites11) [reply service/controller/azureconfig/handler/volumebindingmigration/resource.go:162]: good enought for me, apart from the `panic(err)`

Comment (tuommaki) [reply service/controller/azureconfig/handler/volumebindingmigration/resource.go:154]: You think it should be a handled error despite of built-in templating?

Comment (whites11) [reply service/controller/azureconfig/handler/volumebindingmigration/resource.go:154]: I don't know, will leave that to you.
But the panic feels wrong to me in general.

Comment (tuommaki) [reply service/controller/azureconfig/handler/volumebindingmigration/resource.go:159]: I added extra error handling there.

Comment (whites11) [reply service/controller/azureconfig/handler/volumebindingmigration/resource.go:159]: thx

**Review by whites11 — APPROVED** [review]
LGTM

File: service/controller/azureconfig/handler/volumebindingmigration/resource.go [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+		}
+	}
+
+	return k8sClient.CtrlClient(), nil
+}
+
+func defaultStorageClasses() ([]storagev1.StorageClass, error) {
+	var storageClasses []storagev1.StorageClass
+
+	objs := bytes.Split([]byte(ignition.DefaultStorageClass), []byte("---"))
```
Comment (whites11) [inline service/controller/azureconfig/handler/volumebindingmigration/resource.go:159]: According to https://kubernetes.io/docs/concepts/storage/storage-classes/#volume-binding-mode `AzureFiles` does not support WaitForFirstConsumer. Please let's test if specifica

Comment (tuommaki) [reply service/controller/azureconfig/handler/volumebindingmigration/resource.go:166]: I filter out `AzureFile` storage classes now.

**Comment by tuommaki** [comment]
Works on `ghost` :tm:

**Pull request merged**